Dynamic bandwidth tracking

oxtoacart opened the pull request:
In order to test this, you can locally run a proxy that spits out fake bandwidth tracking stats by building from https://github.com/getlantern/http-proxy-lantern/tree/throttling-ox

oxtoacart commented on lantern-ui/app/js/services.js:
```diff
@@ -90,6 +90,9 @@ angular.module('app.services', [])
         }
       },
+      'bandwidth': function(bandwidth) {
+        console.log('Got bandwidth data: ', bandwidth);
```
@myleshorton Bandwidth data is plumbed through to web browser.  Can you take it from here and bind into the UI as appropriate?

myleshorton replied:
Definitely @oxtoacart thanks!

oxtoacart commented on src/github.com/getlantern/flashlight/ui/service.go:
```diff
@@ -46,7 +45,7 @@ func (s *Service) write() {
 }
 
-func Register(t string, newMessage func() interface{}, helloFn helloFnType) (*Service, error) {
+func Register(t string, helloFn helloFnType) (*Service, error) {
```
Opportunistic cleanup.

myleshorton replied:
Nice!

oxtoacart commented:
The test failure doesn't seem to be isolated to this branch.

oxtoacart commented:
Yeah, lantern_test is failing on devel and master too.

myleshorton commented:
wow really @oxtoacart? I wonder how that happened. Interesting. Does that test live proxying, as in hit real servers?

oxtoacart commented:
@myleshorton Yes, it's hitting real proxies and google.com.

myleshorton commented:
oh interesting -- I wonder if those proxies are failing...

myleshorton commented on src/github.com/getlantern/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,80 @@
+
+// Track updates the bandwith quota information based on the XBQ header in
+// the given response.
+func Track(resp *http.Response) {
```
Can you throw the format we're parsing in a comment here @oxtoacart?

myleshorton replied:
A test would also be ideal but i realize it's late in the game

oxtoacart replied:
I added a comment about the format.

oxtoacart commented:
It looks like it's trying to hit a proxy that doesn't exist.

myleshorton commented on src/github.com/getlantern/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,80 @@
+	}
+	asof := time.Unix(0, asofInt)
+	mutex.Lock()
+	if quota == nil || quota.AsOf.Before(asof) {
```
Since we throw this away when the channel is full, @oxtoacart, I wonder if it might be cleaner to just create a new Quota here and not worry about the locking and reusing an existing one.

myleshorton replied:
I realize we use it in GetQuota, but I think we can also just skip the hello function, which is the only place that's used @oxtoacart

oxtoacart replied:
I think the hello function is important because that's used every time someone opens the UI. Without that, we have to wait for an actual request to be made to display the usage.  The operation inside this lock is really fast, so I wouldn't worry about performance.

myleshorton commented on src/github.com/getlantern/bandwidth/bandwidth.go:
```diff
@@ -0,0 +1,80 @@
+	log = golog.LoggerFor("bandwidth")
+
+	quota *Quota
+	mutex sync.RWMutex
```
Seems like we should be able to skip both of the above to me. If we do really want the hello function, couldn't we just fetch the data from the channel there as well?

oxtoacart replied:
We can't fetch the data from the channel because it's not a buffered channel. Even if we made it a buffered channel, it would still be capped, so at some point the channel would be capturing stale data.

myleshorton commented:
This is overall looking great @oxtoacart, just a few comments.

oxtoacart commented:
@myleshorton Thanks for the review.  I've addressed your comments.  I'd like to leave the GetQuota capability in place as it seems useful for giving up-to-date information to new UI clients.

myleshorton commented:
OK looks good @oxtoacart!
